tests/periph_i2c/main.c

```diff
@@ -20,270 +20,547 @@
         return 1;
     }
-    else if (res == -2) {
-        puts("Error: Init: Unsupported speed value");
+
+    printf("Command: i2c_release(i2c_%i)\n", i2c_dev);
+    res = i2c_release(i2c_dev);
+
+    if (res == I2C_ACK) {
+        printf("Success: i2c_%i successfully released\n", i2c_dev);
+        return 0;
+    }
+    return print_i2c_error(res);
+}
+
+int cmd_i2c_read_reg(int argc, char **argv) {
+    int res;
+    uint8_t addr;
+    uint8_t reg;
+    uint8_t flags = 0;
+    uint8_t data;
+
+    if (check_dev() == 1){
         return 1;
     }
-    else {
-        printf("I2C_%i successfully initialized as master!\n", dev);
-        i2c_dev = dev;
+
+    if (argc < 3 || argc > 4) {
+        INVALID_ARGS;
+        printf("Usage:\n%s ADDR REG [FLAG]\n", argv[0]);
+        return 1;
     }
 
-    return 0;
+    addr = atoi(argv[1]);
+    reg = atoi(argv[2]);
+    if (argc == 4) {
+        flags = atoi(argv[3]);
+    }
+
+    printf("Command: i2c_read_reg(i2c_%i, 0x%02x, 0x%02x)\n", i2c_dev, addr, reg);
+    res = i2c_read_reg(i2c_dev, addr, reg, &data, flags);
+
+
+    if (res == I2C_ACK) {
+        printf("Success: i2c_%i successfully read 1 byte from reg 0x%02x: [",
+            i2c_dev, reg);
+        printf("0x%02x", data);
+        puts("]");
+        return 0;
+    }
+    return print_i2c_error(res);
 }
 
-int cmd_write(int argc, char **argv)
-{
+int cmd_i2c_read_regs(int argc, char **argv) {
     int res;
     uint8_t addr;
-    int length = argc - 2;
+    uint8_t reg;
+    uint8_t flags = 0;
+    int len;
     uint8_t data[BUFSIZE];
 
-    if (i2c_dev < 0) {
-        puts("Error: no I2C device was initialized");
+    if (check_dev() == 1){
         return 1;
     }
-    if (argc < 3) {
-        puts("Error: not enough arguments given");
-        printf("Usage:\n%s: ADDR BYTE0 [BYTE1 [BYTE_n [...]]]\n", argv[0]);
+
+    if (argc < 4 || argc > 5) {
+        INVALID_ARGS;
+        printf("Usage:\n%s ADDR REG LENGTH [FLAG]\n", argv[0]);
         return 1;
     }
 
     addr = atoi(argv[1]);
-    for (int i = 0; i < length; i++) {
-        data[i] = atoi(argv[i + 2]);
+    reg = atoi(argv[2]);
+    len = atoi(argv[3]);
+    if (argc == 5) {
+        flags = atoi(argv[4]);
     }
 
-    if (length == 1) {
-        printf("i2c_write_byte(I2C_%i, 0x%02x, 0x%02x)\n", i2c_dev, addr, data[0]);
-        res = i2c_write_byte(i2c_dev, addr, data[0]);
+    if (len < 1 || len > (int)BUFSIZE) {
+        puts("Error: invalid LENGTH parameter given");
+        return 1;
     }
     else {
-        printf("i2c_write_bytes(I2C_%i, 0x%02x, [", i2c_dev, addr);
-        for (int i = 0; i < length; i++) {
-            printf(", 0x%02x", data[i]);
+        printf("Command: i2c_read_regs(i2c_%i, 0x%02x, 0x%02x, %i)\n", i2c_dev,
+            addr, reg, len);
+        res = i2c_read_regs(i2c_dev, addr, reg, data, len, flags);
+    }
+
+    if (res == I2C_ACK) {
+        printf("Success: i2c_%i successfully read %i bytes from reg 0x%02x: [",
```

Comment:
IMHO its a bit confusing, that the user puts in the address of a register in decimal and its printed in HEX here, or not? At least from what I found `atoi` expects numbers of base 10, but you could use `strtol` with base 16 instead.

Reply:
and btw. maybe this printing could be a helper function? Something like:

```
static inline print_i2c_read(i2c_t dev, uint16_t *reg, uint8_t *buf, size_t len)
{
    assert ((dev > 0) && (len > 0));
    printf("Success: i2c_%i read %i byte(s) ", dev, len);
    if (reg != Null) {
        printf("from reg 0x%02x ", reg);
    }
    printf(": [ ");
    for (unsigned i = 0; i < len; i++) {
        printf("0x%02x ", data[i]);
    }
    puts("]");
}
```

and then be use in `i2c_read_reg(s)/byte(s)`

Reply:
I may just do strtol (legacy used atoi), I know lots of times I like seeing the hex numbers (since most data sheets support that).  With the interface it gets abstracted out though.  Either way I'll improve somehow.  [TODO]